src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1746,7 +1737,7 @@ private void Heapsort(int lo, int hi)
             private void DownHeap(int i, int n, int lo)
             {
                 object d = keys[lo + i - 1];
-                object dt = (items != null) ? items[lo + i - 1] : null;
+                object dt = (items != null) ? items[lo + i - 1] : null!;
```

Comment:
Why use `!` instead of marking dt as nullable?

Reply:
`dt` is only stored back into the array if `items != null`, so it's only used when we didn't introduce a null ourselves.  This goes back to why I chose to make the type `object[]` instead of `object?[]`: we don't actually introduce any nulls that weren't already there.

Reply:
Thanks for the explanation!